ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,665 @@
+  | "bugtool" ->
+      Bugtool
+  | url ->
+      Url (Uri.of_string url)
+
+let ok_none = Ok None
+
+module Status = struct
+  type status_code = Unset | Ok | Error [@@deriving rpcty]
+
+  type t = {status_code: status_code; description: string option}
+  [@@deriving rpcty]
+
+  let status_to_string status =
+    let desc =
+      Option.fold ~none:"" ~some:(fun d -> " " ^ d) status.description
+    in
+    let value =
+      match status.status_code with
+      | Unset ->
+          "Unset"
+      | Ok ->
+          "Ok"
+      | Error ->
+          "Error"
+    in
+    value ^ desc
+end
+
+module SpanContext = struct
+  type t = {trace_id: string; span_id: string} [@@deriving rpcty]
+
+  let to_traceparent t = Printf.sprintf "00-%s-%s-00" t.trace_id t.span_id
+
+  let of_traceparent traceparent =
+    let elements = String.split_on_char '-' traceparent in
+    match elements with
+    | ["00"; trace_id; span_id; "00"] ->
+        Some {trace_id; span_id}
+    | _ ->
+        None
+end
+
+module Span = struct
+  type t = {
+      context: SpanContext.t
+    ; mutable span_kind: SpanKind.t
+    ; status: Status.t
+    ; parent: t option
+    ; name: string
+    ; service_name: string
+    ; begin_time: float
+    ; end_time: float option
+    ; tags: (string * string) list
+  }
+  [@@deriving rpcty]
+
+  let get_context t = t.context
+
+  let generate_id n = String.init n (fun _ -> "0123456789abcdef".[Random.int 16])
+
+  let start ?(tags = []) ~name ~parent ~span_kind ~service_name () =
+    let trace_id =
+      match parent with
+      | None ->
+          generate_id 32
+      | Some span_parent ->
+          span_parent.context.trace_id
+    in
+    let span_id = generate_id 16 in
+    let context : SpanContext.t = {trace_id; span_id} in
+    (* Using gettimeofday over Mtime as it is better for sharing timestamps between the systems *)
+    let begin_time = Unix.gettimeofday () in
+    let end_time = None in
+    let status : Status.t = {status_code= Status.Unset; description= None} in
+    {
+      context
+    ; span_kind
+    ; status
+    ; parent
+    ; name
+    ; service_name
+    ; begin_time
+    ; end_time
+    ; tags
+    }
+
+  let finish ?(tags = []) ~span () =
+    let status = span.status in
+    let tags =
+      if status.status_code = Error then
+        ( Status.status_to_string status
+        , Option.value ~default:"" status.Status.description
+        )
+        :: tags
+      else
+        tags
+    in
+    {span with end_time= Some (Unix.gettimeofday ()); tags= span.tags @ tags}
+
+  let set_span_kind span kind = span.span_kind <- kind
+
+  let set_error span exn =
+    let backtrace = Printexc.get_backtrace () in
+    let description =
+      Some
+        (Printf.sprintf "Error: %s Backtrace: %s" (Printexc.to_string exn)
+           backtrace
+        )
+    in
+    let status_code = Status.Error in
+    match span.status.status_code with
+    | Unset ->
+        {span with status= {status_code; description}}
```

Comment:
Because we do not know if it will be a client or not (The task associated sends a http request) from the start. We can set it to a client if it is used to a set the traceparent header. If there is a better way to do this then that would be an option

Reply:
Resolved by561538ae93abe485bc236f4ab50da0bd3686d11a

Reply:
The alternative to a mutable field in a more functional style is to create a new record with the field updated, as is done here for the status. So I'm questioning why two different methods are used for fields in the same record.

Mutable fields could be convenient if records are stuck in a hash table as you don't have to remove and re-add. Functional updates have other benefits in general.

Reply:
The issue was the updating for this field was deeply embedded in the code and not in the instrumentation, however if we called set before we attach a span in a http header then put it back into context then it should propegate.